src/Gaffer/ScriptNode.cpp

```diff
@@ -607,28 +615,21 @@ void ScriptNode::undo()
 
 	m_currentActionStage = Action::Undo;
 
-		m_undoIterator--;
-		(*m_undoIterator)->undoAction();
-
-	/// \todo It's conceivable that an exception from somewhere in
-	/// Action::undoAction() could prevent this cleanup code from running,
-	/// leaving us in a bad state. This could perhaps be addressed
-	/// by using BOOST_SCOPE_EXIT. The most likely cause of such an
-	/// exception would be in an errant slot connected to a signal
-	/// triggered by the action performed. However, currently most
-	/// python slot callers suppress python exceptions (printing
-	/// them to the shell), so it's not even straightforward to
-	/// write a test case for this potential problem. It could be
-	/// argued that we shouldn't be suppressing exceptions in slots,
-	/// but if we don't then well-behaved (and perhaps crucial) slots
-	/// might not get called when badly behaved slots mess up. It seems
-	/// best to simply report errors as we do, and allow the well behaved
-	/// slots to have their turn - we might even want to extend this
-	/// behaviour to the c++ slots.
-	m_currentActionStage = Action::Invalid;
+	try
+	{
+		// NOTE : Decrement the undo iterator if undoAction() completes without throwing an exception
```

Comment:
If I understand correctly, this means we'll be stuck here and will never be able to undo past this point? What's the rationale for that? Devil's advocate : might it be better to allow the user to move to the next undo in case it helps them recover the state they want?

Reply:
that wont work for the animation stuff as each undo relies on any previous undos successfully completing. Also this mirrors the redo behaviour where the iterator is incremented after the redo completes successfully